gen/internals.py

```diff
@@ -418,49 +649,19 @@ def resolve_configuration(sources: list, targets: list, user_arguments: dict):
         validate += source.validate
 
     # Use setters to caluclate every required parameter
-    arguments = DFSArgumentCalculator(setters, validate).calculate(targets)
-
-    # Validate all new / calculated arguments are strings.
-    validate_arguments_strings(arguments)
-
-    log.info("Final arguments:" + json_prettyprint(arguments))
-
-    # TODO(cmaloney) Give each config target the values for all it's parameters that were hit as
-    # well as any parameters that led to those parameters.
-    return arguments
-
-
-def validate_configuration(sources: list, targets: list, user_arguments: dict):
-    try:
-        resolve_configuration(sources, targets, user_arguments)
-        return {'status': 'ok'}
-    except ValidationError as ex:
-        messages = {}
-
-        # Defer multi-key validation errors and noramlize them to be single-key
-        # ones. The multi-key is always less important than the single-key
-        # messages which is why single-key messages we never overwrite.
-        # TODO(cmaloney): Teach the whole stack to be able to handle multi-key
-        # validation errors.
-        to_do = dict()
-        for key, msg in ex.errors.items():
-            if isinstance(key, frozenset):
-                to_do[key] = msg
-                continue
-
-            assert isinstance(key, str)
-            messages[key] = {'message': msg}
-
-        for keys, msg in to_do.items():
-            for name in keys:
-                # Skip ones we already have one message for.
-                if name in messages:
-                    continue
+    resolver = Resolver(setters, validate, targets)
+    resolver.resolve()
```

Comment:
Talked about this in my above comment.

Practically, I think it makes the code easier to understand / reason about rather than having the one uber function. There is a setup that builds up the state, then a call to a recursive function which uses all that common state to do it's work. Unresolved Resolver() is not an external API. Having a constructor / __init__ function do a lot of work is a bit odd unless it's a Resource Acquisition Is Initialization function (Think a class that holds an fd and wants to guarantee if you have an instance of that class, you have an open fd, and when that class destructs, the fd will always be closed)